### PR TITLE
feat(typescript-estree): always return parserServices

### PR DESCRIPTION
**Pull request by bradzacher**
BREAKING CHANGE:

previously plugins expected that `parserServices` would be `undefined` if full type information wasn't available.

Splitting these changes out of #688 to focus that PR.
The motivation behind this is that this will allow rules to consume diagnostics without needing the user to provide `project`, which slows down linting.

### CODE REVIEW

**Comment by armano2** [comment]
@bradzacher are you planning to resolve conflicts or do you want me to pick it up?

**Comment by bradzacher** [comment]
I was planning on getting back to it eventually, but it's been at a lower priority, because 3.0.0 was so far out 😅.

If you get the gist of what I was intending, then feel free to pick this up and get it ready for `3.0.0`. I'd appreciate it 😄.

One important note: `no-unused-vars-experimental` will not require the complete type information with this change, so you should update that rule as part of this PR.

File: packages/typescript-estree/tests/lib/parse.ts [hunk 1]
```diff
@@ -289,6 +294,8 @@ describe('parse()', () => {
         }
 
         if (!shouldThrow) {
+          expect(result.services.program).toBeDefined();
+          result.services.program = {}; // reduce noise
```
Comment (armano2) [inline packages/typescript-estree/tests/lib/parse.ts]: we do not want to output all of internal typescript functions to snapshot here

File: packages/typescript-estree/src/parser.ts [hunk 2]
```diff
@@ -244,13 +239,10 @@ function applyParserOptionsToExtra(options: TSESTreeOptions): void {
...
-   *
-   * NOTE: For backwards compatibility we also preserve node maps in the case where `project` is set,
-   * and `preserveNodeMaps` is not explicitly set to anything.
    */
   extra.preserveNodeMaps =
     typeof options.preserveNodeMaps === 'boolean' && options.preserveNodeMaps;
-  if (options.preserveNodeMaps === undefined && extra.projects.length > 0) {
+  if (options.preserveNodeMaps === undefined) {
     extra.preserveNodeMaps = true;
   }
```
Comment (bradzacher) [inline packages/typescript-estree/src/parser.ts]: This should probably just be

```ts
if (typeof options.preserveNodeMaps === 'boolean') {
  extra.preserveNodeMaps = options.preserveNodeMaps;
}
```

File: packages/typescript-estree/src/parser.ts [hunk 3]
```diff
@@ -296,6 +288,7 @@ interface ParseAndGenerateServicesResult<T extends TSESTreeOptions> {
 // Public
 //------------------------------------------------------------------------------
 
+// note - cannot migrate this to an import statement because it will make TSC copy the package.json to the dist folder
 const version: string = require('../package.json').version;
```
Comment (bradzacher) [inline packages/typescript-estree/src/parser.ts]: unrelated:
should probably just add this as an internal lint rule.

File: packages/experimental-utils/src/ts-eslint/Rule.ts [hunk 4]
```diff
@@ -185,7 +185,7 @@ interface RuleContext<
   /**
    * An object containing parser-provided services for rules
    */
-  parserServices?: ParserServices;
+  parserServices: ParserServices;
```
Comment (bradzacher) [inline packages/experimental-utils/src/ts-eslint/Rule.ts]: question:
considering these are more general types for eslint, should we keep this as optional?
that way you still have to explicitly handle the case that someone uses a different parser.

File: packages/eslint-plugin/src/rules/no-unused-vars-experimental.ts [hunk 5]
```diff
@@ -68,7 +68,7 @@ export default util.createRule<Options, MessageIds>({
     },
   ],
   create(context, [userOptions]) {
-    const parserServices = util.getParserServices(context);
+    const parserServices = util.getParserServices(context, true);
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-unused-vars-experimental.ts]: I can't comment on the line above, but you can remove `requiresTypeChecking` from this rule's metadata now.

Comment (armano2) [reply packages/eslint-plugin/src/rules/no-unused-vars-experimental.ts]: its not that simple, unit test has to be updated first :>

File: packages/eslint-plugin/tests/docs.test.ts [hunk 6]
```diff
@@ -85,9 +85,14 @@ describe('Validating rule metadata', () => {
         // not perfect but should be good enough
         const ruleFileContents = fs.readFileSync(
           path.resolve(__dirname, `../src/rules/${ruleName}.ts`),
+          'utf-8',
         );
 
-        expect(ruleFileContents.includes('getParserServices')).toEqual(
+        const requiresTypeChecking = /getParserServices(\(\s*[a-z]+\s*)\)/.test(
+          ruleFileContents,
+        );
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]: this should also catch `getParserServices(context, true)`

Comment (armano2) [reply packages/eslint-plugin/tests/docs.test.ts]: it should also catch
```
getParserServices(context, false)
```

but yes, added :>

**Pull request merged**